### PR TITLE
branch: Fix delete branch api in client and mock client

### PR DESCRIPTION
**Pull request by aditya-nambiar**
## Summary by CodeRabbit

- **Bug Fixes**
	- Improved the `delete_branch` method to prevent deletion of the main branch and ensure branch consistency.
- **Tests**
	- Added new tests for branch deletion functionality, including edge cases.
	- Updated branch cloning error tests for more precise exception handling.
	- Marked a test as slow to improve test suite performance.
- **Documentation**
	- Updated the changelog for version 1.2.3.
- **Chores**
	- Updated the project version to 1.2.3 in `pyproject.toml`.

### CODE REVIEW

File: fennel/client/client.py [hunk 1]
```diff
@@ -383,8 +383,17 @@ def delete_branch(self, name: str):
         name (str): The name of the branch to delete.
 
         """
+        if name == _MAIN_BRANCH:
+            raise Exception("Cannot delete the main branch.")
+
+        cur_branch = self._branch
+        self.checkout(name)
```
Comment (bothra90) [inline fennel/client/client.py:390]: we don't need this, right?

File: fennel/client/client.py [hunk 2]
```diff
@@ -383,8 +383,17 @@ def delete_branch(self, name: str):
...
         response = self._post_json(f"{V1_API}/delete", {})
-        self.checkout(_MAIN_BRANCH)
+        if name == cur_branch:
+            print("Deleting current branch, checking out to main branch")
+            self.checkout(_MAIN_BRANCH)
+        else:
+            self.checkout(cur_branch)
```
Comment (bothra90) [inline fennel/client/client.py:396]: we can skip this too

**Review by bothra90 — APPROVED** [review]
LGTM. let's add a commit message for the change in behavior

File: fennel/client/client.py (hunk 1 above)
Comment (aditya-nambiar) [reply fennel/client/client.py:390]: We checkout the 'name' branch delete it and then checkout the current branch

**Review by petrpan26 — COMMENTED** [review]
LGTM, some coderabbit comment seems reasonable, feel free to resolve what reasonable !
